context/src/main/java/io/micronaut/runtime/beans/DefaultIntrospectionBeanMapHandler.java

```diff
@@ -0,0 +1,305 @@
+import io.micronaut.core.beans.BeanProperty;
+import io.micronaut.core.beans.UnsafeBeanProperty;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.reflect.ReflectionUtils;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.util.CollectionUtils;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * BeanMapper implementation that uses introspections.
+ *
+ * @since 4.1.0
+ */
+@Singleton
+@Internal
+final class DefaultIntrospectionBeanMapHandler implements IntrospectionBeanMapHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultIntrospectionBeanMapHandler.class);
+    private static final String PREFIX_UNABLE_BIND_PROPERTY = "Unable bind property [";
+    private static final String MSG_TO_TARGET_BEAN = "] to target bean [";
+    private static final String PREFIX_NON_NULL_ARG = "Non-null constructor argument specified as null: ";
+    private final BeanIntrospector beanIntrospector;
+    private final ConversionService conversionService;
+
+    /**
+     * Constructor that specifies a specific introspector.
+     *
+     * @param beanIntrospector  The introspector
+     * @param conversionService The conversion service
+     */
+    DefaultIntrospectionBeanMapHandler(@NonNull BeanIntrospector beanIntrospector, ConversionService conversionService) {
+        this.beanIntrospector = Objects.requireNonNull(beanIntrospector, "BeanIntrospector cannot be null!");
+        this.conversionService = conversionService;
+    }
+
+    /**
+     * Default constructor.
+     * @param conversionService  The conversion service.
+     */
+    @Inject
+    DefaultIntrospectionBeanMapHandler(ConversionService conversionService) {
+        this(BeanIntrospector.SHARED, conversionService);
+    }
+
+    @Override
+    public <I, O> O map(I input, O output, BeanMapper.MapStrategy mapStrategy, BeanIntrospection<I> left, BeanIntrospection<O> right) {
+        if (mapStrategy == null) {
+            mapStrategy = BeanMapper.MapStrategy.DEFAULT;
+        }
+        Collection<BeanProperty<I, Object>> beanProperties = left.getBeanProperties();
+        BeanMapper.MapStrategy.ConflictStrategy conflictStrategy = mapStrategy.conflictStrategy();
+        if (conflictStrategy == null || conflictStrategy == BeanMapper.MapStrategy.ConflictStrategy.ERROR) {
+
+                for (BeanProperty<I, Object> leftProperty : beanProperties) {
+                    try {
+                        BeanProperty<O, Object> rightProperty =
+                                right.getRequiredProperty(leftProperty.getName(), leftProperty.getType());
+                        if (!rightProperty.isReadOnly()) {
+                            ((UnsafeBeanProperty<O, Object>) rightProperty).setUnsafe(output, leftProperty.get(input));
```

Comment:
Do we support immutable beans?

Reply:
Those that can be mutated by creating a new instance like records

Reply:
yes records are supported

Reply:
I think you should use `withValueUnsafe`